examples/Custom_Hooks/probability_of_improvement.py

```diff
@@ -0,0 +1,149 @@
+from baybe.targets import NumericalTarget
+from baybe.utils.basic import register_hooks
+from baybe.utils.dataframe import add_fake_results, to_tensor
+
+### Setup
+
+# We start by initialize the global poi_list where we will store the probability of improvement from each recommend iteration
+
+poi_list = []
+
+
+# Then, we define the hook that calculates the probability of improvement based on the given recommender, search space and measurements:
+
+
+def check_probability_of_improvement(
+    self: BotorchRecommender,
+    searchspace: SearchSpace,
+    objective: Objective | None = None,
+    measurements: pd.DataFrame | None = None,
+):
+    """Calculate and store the probability of improvement in poi_list."""
+    if searchspace.type == SearchSpaceType.DISCRETE:
+        train_x = searchspace.transform(measurements)
+        train_y = objective.transform(measurements)
+        ac_f = ProbabilityOfImprovement()
+        botorch_acf = ac_f.to_botorch(
+            self.surrogate_model, searchspace, train_x, train_y
+        )
+        comp_rep_tensor = to_tensor(searchspace.discrete.comp_rep).unsqueeze(1)
+        poi = botorch_acf(comp_rep_tensor)
+        poi_list.append(poi)
+    else:
+        raise TypeError(
+            f"{searchspace.type} search spaces are not supported yet. "
+            f"Currently only DISCRETE search spaces are accepted."
+        )
+
+
+# Additionally, we define a function that plots the poi_list after all recommend  iterations:
+
+
+def plot_poi(poi_list: list):
+    """Plot the probability of improvement."""
+    num_plots = len(poi_list)
+    fig, axs = plt.subplots(num_plots, 1, figsize=(8, 6 * num_plots))
+
+    # Plot each POI tensor separately
+    for i, p in enumerate(poi_list):
+        poi_np = p.detach().numpy()
+        sns.histplot(
+            poi_np,
+            kde=True,
+            stat="density",
+            color="mediumaquamarine",
+            bins=20,
+            edgecolor="black",
+            linewidth=0.5,
+            ax=axs[i],
+        )
+        axs[i].set_xlabel(f"Iteration {i+1}: Probability of Improvement")
+        axs[i].set_ylabel("Search space Points")
+
+    fig.subplots_adjust(hspace=0.5)
+    plt.suptitle("Investigate the probability of improvement")
+    plt.show()
+
+
+### Monkeypatching
+
+# Next, we create our recommender and monkeypatch its `recommend` method:
+
+my_recommender = TwoPhaseMetaRecommender(
+    initial_recommender=RandomRecommender(),
+    recommender=BotorchRecommender(
+        surrogate_model=GaussianProcessSurrogate(),
+        allow_repeated_recommendations=True,
+        allow_recommending_already_measured=True,
+    ),
+)
+my_recommender.recommender.recommend = MethodType(
+    register_hooks(
+        BotorchRecommender.recommend,
+        post_hooks=[check_probability_of_improvement],
+    ),
+    my_recommender.recommender,
+)
+
+
+### Triggering the Hooks
+
+# We setup the other objects to trigger the hook:
+
+parameters = [
```

Comment:
please set this up with a more reasonable example ( a function that will genereate a disribution of PIs and not some trivial few-bin values)

I would alsos uggest to include `SMOKE_TEST` as in many other examples so you can develop with speed but have better pictures if run from a production

Reply:
`POINTS_PER_DIM` is not affected by SMOKE_TEST which is unusual (see all other examples)